packages/material-ui/src/FormControl/FormControl.js

```diff
@@ -49,114 +48,121 @@ export const styles = {
  *
  * ⚠️ Only one input can be used within a FormControl.
  */
-class FormControl extends React.Component {
-  static getDerivedStateFromProps(props, state) {
-    if (props.disabled && state.focused) {
-      return { focused: false };
-    }
-    return null;
-  }
+const FormControl = React.forwardRef(function FormControl(props, ref) {
+  const {
+    children,
+    classes,
+    className,
+    component: Component,
+    disabled,
+    error,
+    fullWidth,
+    margin,
+    required,
+    variant,
+    ...other
+  } = props;
+  const [adornedStart] = React.useState(() => {
+    // We need to iterate through the children and find the Input in order
+    // to fully support server-side rendering.
+    let initialAdornedStart = false;
```

Comment:
I'm not too happy about setting initial state like this. There must be a simpler/nicer way. I'll probably pull this out into a function for readability.

Reply:
https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily

Sounds like this is perfectly fine. I'm not a big fan of using state like this either. But it doesn't fit "lazy ref initialization" because we need it immediately (do we?). It's probably the best pattern we have right now.

Reply:
Yeah, the most noticeable difference is it loops through it’s children twice instead of once.

Reply:
I don't think that's an issue. Better keep different things separated than prematurely combine unrelated things unless this was specifically added to address some performance concern. Just use Big-O notation and looping twice becomes a pretty `O(n)` :smile: